File: cucumber-scalatest/src/main/java/io/cucumber/scalatest/CucumberOptions.java
package io.cucumber.scalatest;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Configuration annotation for Cucumber tests in ScalaTest.
 *
 * Use this annotation to configure Cucumber options for your test suite.
 *
 * Example:
 * <pre>
 * &#64;CucumberOptions(
 *   glue = {"com.example.stepdefinitions"},
 *   features = {"classpath:features"},
 *   plugin = {"pretty"}
 * )
 * class RunCucumberTest extends CucumberSuite
 * </pre>
 */
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.TYPE})
public @interface CucumberOptions {

    /**
     * @return the package(s) containing the glue code (step definitions and hooks).
     */
    String[] glue() default {};

    /**
     * @return the paths to the feature file(s) or directory(ies) on the classpath.
     */
    String[] features() default {};

    /**
     * @return output plugins to use
     */
    String[] plugin() default {};

    /**
     * @return only run scenarios tagged with tags matching this expression
     */
    String tags() default "";

    /**
     * @return true if glue code should be executed in dry run mode
     */
    boolean dryRun() default false;

    /**
     * @return true if output should not use ANSI color escape codes
     */
    boolean monochrome() default false;

    /**
     * @return only run scenarios matching this name
     */
    String[] name() default {};

    /**
     * @return snippets should use this snippet type
     */
    String snippets() default "UNDERSCORE";

    /**
     * @return the object factory class
     */
    Class<?> objectFactory() default Object.class;

    /**
     * @return true if results should be published
     */
    boolean publish() default false;
}

File: cucumber-scalatest/src/main/scala/io/cucumber/scalatest/CucumberOptions.scala
package io.cucumber.scalatest

/** Configuration for Cucumber tests.
  *
  * <p>
  * <strong><u>Features</u></strong>
  * <p>
  * A list of features paths
  * <p> A feature path is constructed as
  * {@code [ PATH[.feature[:LINE]*] | URI[.feature[:LINE]*] | @PATH ]} <p>
  * Examples:
  * <ul>
  *   <li>{@code src/test/resources/features} -- All features in the {@code src/test/resources/features} directory</li>
  *   <li>{@code classpath:com/example/application} -- All features in the {@code com.example.application} package</li>
  *   <li>{@code in-memory:/features} -- All features in the {@code /features} directory on an in memory file system
  *   supported by {@link java.nio.file.FileSystems}</li>
  *   <li>{@code src/test/resources/features/example.feature:42} -- The scenario or example at line 42 in the example
  *   feature file</li>
  * </ul>
  * <p>
  * When no feature path is provided, Cucumber will use the package of the suite class.
  * For example, if the suite class is {@code com.example.RunCucumber} then features are assumed to be located in
  * {@code classpath:com/example}.
  *
  * <p>
  * <strong><u>Glue</u></strong>
  * <p>
  * Package to load glue code (step definitions, hooks and plugins) from. E.g: {@code com.example.app}
  * <p>
  * When no glue is provided, Cucumber will use the package of the suite class.
  * For example, if the suite class is {@code com.example.RunCucumber} then glue is assumed to be located in {@code com.example}.
  *
  * <p>
  * <strong><u>Plugins</u></strong>
  * <p>
  * Register plugins. Built-in plugin types: {@code junit}, {@code html}, {@code pretty}, {@code progress}, {@code json},
  * {@code usage}, {@code unused}, {@code rerun}, {@code testng}.
  * <p>
  * Can also be a fully qualified class name, allowing registration of 3rd party plugins.
  * <p>
  * Plugins can be provided with an argument. For example {@code json:target/cucumber-report.json}
  *
  * <p>
  * <strong><u>Tags</u></strong>
  * <p>
  * Only run scenarios tagged with tags matching <a href="https://github.com/cucumber/tag-expressions">Tag Expression</a>.
  * <p>
  * For example {@code "@smoke and not @fast"}.
  *
  * @param features A list of features paths
  * @param glue Package to load glue code (step definitions, hooks and plugins) from. E.g: {@code com.example.app}
  * @param plugin Plugins
  * @param tags Tags
  */
case class CucumberOptions(
    features: List[String] = List.empty,
    glue: List[String] = List.empty,
    plugin: List[String] = List.empty,
    tags: Option[String] = None
)
